File: PM/TimeUsage/app/src/main/java/com/dominik/time/MainActivity.java
package com.dominik.time;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity
{
    private static final String FRONT_TIME = "FRONT_TIME";
    private static final String TOTAL_TIME = "TOTAL_TIME";
    private static final String RUN_COUNT = "RUN_COUNT";
    private static final String TICK = "TICK";
    private boolean isVisible = false;
    private int runCount;
    private int totalTime;
    private int frontTime;
    BroadcastReceiver receiver;

    private void showData()
    {
        ((TextView) findViewById(R.id.totalTime)).setText(String.valueOf(totalTime));
        ((TextView) findViewById(R.id.frontTime)).setText(String.valueOf(frontTime));
        ((TextView) findViewById(R.id.runCount)).setText(String.valueOf(runCount));
    }

    private void retrieveNumbers(Bundle savedInstanceState)
    {
        runCount = savedInstanceState.getInt(RUN_COUNT);
        totalTime = savedInstanceState.getInt(TOTAL_TIME);
        frontTime = savedInstanceState.getInt(FRONT_TIME);
    }

    private void retrieveNumbers(SharedPreferences preferences)
    {
        runCount = preferences.getInt(RUN_COUNT, 0) + 1;
        totalTime = preferences.getInt(TOTAL_TIME, 0);
        frontTime = preferences.getInt(FRONT_TIME, 0);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        if(receiver == null)
        {
            receiver = new BroadcastReceiver() {
                @Override
                public void onReceive(Context context, Intent intent) {
                    frontTime += (isVisible ? 1 : 0);
                    totalTime++;
                    showData();
                }
            };
            LocalBroadcastManager.getInstance(this).registerReceiver(receiver, new IntentFilter(TICK));
            SleepService.startTicking(this);
        }

        if(savedInstanceState != null)
            retrieveNumbers(savedInstanceState);
        else
            retrieveNumbers(getPreferences(MODE_PRIVATE));
        setContentView(R.layout.activity_main);
        showData();
    }

    @Override
    public void onRestoreInstanceState(Bundle savedInstanceState)
    {
        super.onRestoreInstanceState(savedInstanceState);
    }

    @Override
    public void onSaveInstanceState(Bundle outState)
    {
        super.onSaveInstanceState(outState);
        outState.putInt(FRONT_TIME, frontTime);
        outState.putInt(TOTAL_TIME, totalTime);
        outState.putInt(RUN_COUNT, runCount);
    }

    @Override
    protected void onStart()
    {
        super.onStart();
    }

    @Override
    protected void onResume()
    {
        super.onResume();
        isVisible = true;
    }

    @Override
    protected void onPause()
    {
        super.onPause();
    }

    @Override
    protected void onStop()
    {
        super.onStop();
        isVisible = false;
    }

    @Override
    protected void onDestroy()
    {
        super.onDestroy();
        SharedPreferences.Editor editor = getPreferences(MODE_PRIVATE).edit();
        editor.putInt(RUN_COUNT, runCount);
        editor.putInt(TOTAL_TIME, totalTime);
        editor.putInt(FRONT_TIME, frontTime);
        editor.apply();
        LocalBroadcastManager.getInstance(this).unregisterReceiver(receiver);
        SleepService.stopTicking(this);
    }
}

File: PM/UrlChecker/app/src/main/java/com/dominik/checker/MainActivity.kt
package com.dominik.checker

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Color
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v4.content.LocalBroadcastManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.dominik.checker.CheckerService.Companion.BROADCAST
import com.dominik.checker.CheckerService.Companion.LIST


const val TAG = "MainActivity"


class MainActivity : AppCompatActivity() {

    private val array = ArrayList<String>()
    private lateinit var adapter: UrlAdapter
    lateinit var list: ListView
    lateinit var editText: EditText
    lateinit var addButton: Button
    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            if (intent == null)
                return
            when (intent.action) {
                BROADCAST -> handleBroadcast(intent)
                else -> Log.w(TAG, "Strange intent with action: '${intent.action}' was received")
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        list = findViewById(R.id.list)
        editText = findViewById(R.id.edit_text)
        addButton = findViewById(R.id.add_button)

        adapter = UrlAdapter(this, R.id.list_item, array)
        list.adapter = adapter
        LocalBroadcastManager.getInstance(this).registerReceiver(receiver, IntentFilter(BROADCAST))

        CheckerService.startActionBroadcast(this)
    }

    fun handleBroadcast(intent: Intent) {
        val arrayList = intent.getSerializableExtra(LIST) as ArrayList<*>
        array.clear();
        arrayList.forEach { array.add(it as String) }
        adapter.notifyDataSetChanged()
    }

    fun addButtonClick(view: View?) {
        val url = editText.text.toString()
        editText.text.clear()
        CheckerService.startActionAdd(this, url)
        if(!array.contains(url))
        {
            array.add(url)
            adapter.notifyDataSetChanged()
        }
    }

    companion object {
        class UrlAdapter(context: Context?, resource: Int, list: ArrayList<String>?) : ArrayAdapter<String>(context, resource, list) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
                val item = getItem(position)
                val view = convertView
                        ?: LayoutInflater.from(context).inflate(R.layout.list_item, null)

                val url = view.findViewById<TextView>(R.id.url_name)
                val button = view.findViewById<ImageButton>(R.id.remove_button)
                url.text = item
                button.setOnClickListener {
                    CheckerService.startActionDelete(context, item)
                }
                view.setBackgroundColor(Color.parseColor(if (position % 2 == 0) "#FFFFFF" else "#DCE2DF"))
                return view
            }
        }

    }
}
